src/v/cluster/cloud_metadata/offsets_uploader.cc

```diff
@@ -47,35 +48,44 @@ ss::future<offsets_upload_result> offsets_uploader::upload(
   const model::cluster_uuid& uuid,
   const model::ntp& ntp,
   const cluster_metadata_id& meta_id,
-  retry_chain_node& retry_node) {
+  retry_chain_node& parent_node) {
     vlog(clusterlog.debug, "Requested to upload offsets from {}", ntp);
     auto holder = _gate.hold();
     _as.check();
-    auto snap_res = co_await _group_manager.local().snapshot_groups(ntp);
+    auto snap_res = co_await _group_manager.local().snapshot_groups(
+      ntp,
+      config::shard_local_cfg()
+        .cloud_storage_cluster_metadata_num_consumer_groups_per_upload());
     if (snap_res.has_error()) {
         co_return snap_res.error();
     }
-    auto& snap = snap_res.value();
-    auto buf = serde::to_iobuf(std::move(snap));
-    auto remote_key = offsets_snapshot_key(uuid, meta_id, ntp.tp.partition);
-
+    size_t snap_idx = 0;
+    auto& snaps = snap_res.value();
     offsets_upload_paths paths;
-    try {
-        auto upload_res = co_await _remote.local().upload_object(
-          _bucket, remote_key, std::move(buf), retry_node);
-        if (upload_res == cloud_storage::upload_result::success) {
-            paths.paths.emplace_back(remote_key().c_str());
-            co_return paths;
-        }
-    } catch (...) {
-        auto eptr = std::current_exception();
-        if (ssx::is_shutdown_exception(eptr)) {
-            vlog(clusterlog.debug, "Shutdown while uploading offsets");
-        } else {
-            vlog(clusterlog.debug, "Error while uploading offsets: {}", eptr);
+    for (auto& snap : snaps) {
+        retry_chain_node retry_node(&parent_node);
+        auto buf = serde::to_iobuf(std::move(snap));
+        auto remote_key = offsets_snapshot_key(
+          uuid, meta_id, ntp.tp.partition, snap_idx++);
+
+        try {
+            auto upload_res = co_await _remote.local().upload_object(
+              _bucket, remote_key, std::move(buf), retry_node);
+            if (upload_res == cloud_storage::upload_result::success) {
+                paths.paths.emplace_back(remote_key().c_str());
+            }
+        } catch (...) {
+            auto eptr = std::current_exception();
+            if (ssx::is_shutdown_exception(eptr)) {
+                vlog(clusterlog.debug, "Shutdown while uploading offsets");
+            } else {
+                vlog(
+                  clusterlog.debug, "Error while uploading offsets: {}", eptr);
+            }
+            co_return error_outcome::upload_failed;
```

Comment:
is there any difference in behavior to take into consideration now that we can have a partial upload failure? here, or in restore to identify partial upload?

Reply:
Not really -- if there is a failure to upload, an error is returned and the paths aren't passed back to the manifest. In that case, we do not update the manifest

https://github.com/redpanda-data/redpanda/blob/8132430201137433771aeb41fd0583d3c9915f0c/src/v/cluster/cloud_metadata/uploader.cc#L173

This does mean that the "snapshot" may not reflect the same round of metadata uploads, but at least it should still represent the partition completely at a given (stale) point in time